src/fsharp/TypeChecker.fs

```diff
@@ -5641,12 +5641,14 @@ and TcExprUndelayed cenv overallTy env tpenv (expr: SynExpr) =
         TcRecdExpr cenv overallTy env tpenv (inherits,optOrigExpr,flds,mWholeExpr)
 
     | SynExpr.While (spWhile,e1,e2,m) ->
+        let env = ClearContext env
```

Comment:
I expect you'll have to clear on all recursive calls.  I think it may be more accurate to just pass the ctxt as an explicit parameter to TcExpr?